module/zfs/spa_config.c

```diff
@@ -66,7 +66,6 @@ static uint64_t spa_config_generation = 1;
  * userland pools when doing testing.
  */
 char *spa_config_path = ZPOOL_CACHE;
-int zfs_autoimport_disable = 1;
```

Comment:
When removing this we want to make sure to preserve the existing behavior.  FreeBSD behaves differently than Linux in this regard.

Traditionally on Linux when loading a kernel module it's not expected to have any side effects like accessing a configuration file, so this behavior was disabled by default.  Since we do want FreeBSD to behave differently why don't we do something like this:

```diff
diff --git a/module/zfs/spa_config.c b/module/zfs/spa_config.c
index 3a65b0bb432..bb130d02bdc 100644
--- a/module/zfs/spa_config.c
+++ b/module/zfs/spa_config.c
@@ -66,7 +66,17 @@ static uint64_t spa_config_generation = 1;
  * userland pools when doing testing.
  */
 char *spa_config_path = ZPOOL_CACHE;
-int zfs_autoimport_disable = 1;
+
+/*
+ * Disabled by default on Linux where the expectation is that loading a
+ * kernel module will not result in side effects such as reading a
+ * configuration file.
+ */
+#if defined(_KERNEL) && defined(__linux__)
+static int spa_config_load_namespace = 0;
+#else
+static int spa_config_load_namespace = 1;
+#endif
 
 /*
  * Called when the module is first loaded, this routine loads the configuration
@@ -85,10 +95,8 @@ spa_config_load(void)
        uint64_t fsize;
        int err;
 
-#ifdef _KERNEL
-       if (zfs_autoimport_disable)
+       if (spa_config_load_namespace)
                return;
-#endif
 
        /*
         * Open the configuration file.
```

Since this has been the default behavior on Linux for many years we can still remove the module option and documentation.

Reply:
Admittedly, I found the module param useful on FreeBSD. Differentiating the 'boot' case, from when I manually load the kernel module (where it is strange to auto-import stuff).

But, your approach makes sense to me.

Reply:
Well we don't have to get rid of the module option at all if we don't want to.  Nor do we need to change the name.  If it's been useful we could just document the different platform defaults.

Reply:
it may be worth emitting a kernel message warning the user that it will be removed some time in the future rather than just yanking it out.

Reply:
is the logic switched around in your diff - it's supposed to *not* open the config file if __linux__ is defined, right?